pkg/ebpf/c/tc.v6egress.bpf.c

```diff
@@ -119,119 +115,109 @@ int handle_egress(struct __sk_buff *skb)
-				l4_dst_port = (((((unsigned short)(l4_tcp_hdr->dest) & 0xFF)) << 8) | (((unsigned short)(l4_tcp_hdr->dest) & 0xFF00) >> 8));
-				break;
-			case IPPROTO_UDP:
-				if (data + sizeof(*ip) + sizeof(*l4_udp_hdr) > data_end) {
-					return BPF_OK;
-				}
-				l4_src_port = (((((unsigned short)(l4_udp_hdr->source) & 0xFF)) << 8) | (((unsigned short)(l4_udp_hdr->source) & 0xFF00) >> 8));
-				l4_dst_port = (((((unsigned short)(l4_udp_hdr->dest) & 0xFF)) << 8) | (((unsigned short)(l4_udp_hdr->dest) & 0xFF00) >> 8));
-				break;
-			case IPPROTO_SCTP:
-				if (data + sizeof(*ip) + sizeof(*l4_sctp_hdr) > data_end) {
-					return BPF_OK;
-				}
-				l4_src_port = (((((unsigned short)(l4_sctp_hdr->source) & 0xFF)) << 8) | (((unsigned short)(l4_sctp_hdr->source) & 0xFF00) >> 8));
-				l4_dst_port = (((((unsigned short)(l4_sctp_hdr->dest) & 0xFF)) << 8) | (((unsigned short)(l4_sctp_hdr->dest) & 0xFF00) >> 8));
-				break;
-		}
-
-		trie_key.prefix_len = 128;
+	trie_key.prefix_len = 128;
+		
         //Fill the IP Key to be used for lookup
         for (int i=0; i<16; i++){
               trie_key.ip[i] = ip->daddr.in6_u.u6_addr8[i];
-              src_ip[i] = ip->saddr.in6_u.u6_addr8[i];
-              flow_key.src_ip[i] = ip->saddr.in6_u.u6_addr8[i];
-              flow_key.dest_ip[i] = ip->daddr.in6_u.u6_addr8[i];
-              reverse_flow_key.src_ip[i] = ip->daddr.in6_u.u6_addr8[i];
-              reverse_flow_key.dest_ip[i] = ip->saddr.in6_u.u6_addr8[i];
-              evt.src_ip[i] = ip->saddr.in6_u.u6_addr8[i];
-              evt.dest_ip[i] = ip->daddr.in6_u.u6_addr8[i];
         }
 
-		//Check for the an existing flow in the conntrack table
-		flow_key.src_port = l4_src_port;
-		flow_key.dest_port = l4_dst_port;
-		flow_key.protocol = ip->nexthdr;
-
-
-		//Check if it's an existing flow
-		flow_val = (struct conntrack_value *)bpf_map_lookup_elem(&aws_conntrack_map, &flow_key);
-		if (flow_val != NULL &&(flow_val->val[12] == src_ip[12] && flow_val->val[13] == src_ip[13]
-					&& flow_val->val[14] == src_ip[14] && flow_val->val[15] == src_ip[15])) {
-			return BPF_OK;
-		}
-
-		evt.src_port = flow_key.src_port;
-		evt.dest_port = flow_key.dest_port;
-		evt.protocol = flow_key.protocol;
-
-		//Check for the reverse flow entry in the conntrack table
-		reverse_flow_key.src_port = l4_dst_port;
-		reverse_flow_key.dest_port = l4_src_port;
-		reverse_flow_key.protocol = ip->nexthdr;
+	//Check for the an existing flow in the conntrack table
+	flow_key.saddr = ip->saddr;
+	flow_key.daddr = ip->daddr;
+	flow_key.src_port = l4_src_port;
+	flow_key.dest_port = l4_dst_port;
+	flow_key.protocol = ip->nexthdr;
+
+	//Check if it's an existing flow
+	flow_val = (struct conntrack_value *)bpf_map_lookup_elem(&aws_conntrack_map, &flow_key);
+	if (flow_val != NULL && (flow_val->addr.in6_u.u6_addr8[12] == flow_key.saddr.in6_u.u6_addr8[12] 
```

Comment:
can we use `__builtin_memcmp`?

Reply:
Yes I did try using this - 

```
#define IN6_ARE_ADDR_EQUAL(a, b)			\
    (__builtin_memcmp(&(a)->in6_u.u6_addr8[0], &(b)->in6_u.u6_addr8[0], sizeof(struct in6_addr)) == 0)
```

But the compiler wasn't happy hence have kept the #define in the code which we can revisit.

Reply:
thanks!